src/indexer.rs

```diff
@@ -94,3 +100,159 @@ impl<'a> Binding for Progress<'a> {
 )]
 #[allow(dead_code)]
 pub type TransportProgress<'a> = IndexerProgress<'a>;
+
+/// A stream to write and index a packfile
+///
+/// This is equivalent to [`crate::OdbPackwriter`], but allows to store the pack
+/// and index at an arbitrary path. It also does not require access to an object
+/// database if, and only if, the pack file is self-contained (i.e. not "thin").
+pub struct Indexer<'odb> {
+    raw: *mut raw::git_indexer,
+    progress: MaybeUninit<raw::git_indexer_progress>,
+    progress_payload_ptr: *mut OdbPackwriterCb<'odb>,
+}
+
+impl<'a> Indexer<'a> {
+    /// Create a new indexer
+    ///
+    /// The [`Odb`] is used to resolve base objects when fixing thin packs. It
+    /// can be `None` if no thin pack is expected, in which case missing bases
+    /// will result in an error.
+    ///
+    /// `mode` is the permissions to use for the output files, use `0` for defaults.
+    ///
+    /// If `verify` is `false`, the indexer will bypass object connectivity checks.
+    pub fn new(odb: Option<&Odb<'a>>, path: &Path, mode: u32, verify: bool) -> Result<Self, Error> {
+        let path = path.into_c_string()?;
+
+        let odb = odb.map(Binding::raw).unwrap_or_else(ptr::null_mut);
+
+        let mut out = ptr::null_mut();
+        let progress = MaybeUninit::uninit();
```

Comment:
I'm having trouble seeing where `progress` is ever initialized. I see it created here, and then a few calls to `as_mut_ptr`, but I don't see where it gets initialized before `as_mut_ptr` is called. Can you help me understand how this works?

Reply:
I'm not sure, actually, `OdbPackwriter` does the same. I _assumed_ it works because `MaybeUninit<T>` has the same size and alignment as `T`, so the C side can just write to it (i.e. it is initialized after the first time we pass `as_mut_ptr` over to `libgit2`).

I don't know why this would be preferable over deriving `Default` for `git_indexer_progress`, and casting to a raw pointer where needed. If I made this change, would you want me to do it for `OdbPackwriter` as well?

Reply:
Yea, it looks like OdbPackwriter should also be updated. From what I can tell, the C side is writing to uninitialized memory which is UB to my understanding.